deckhouse-controller/pkg/controller/module-controllers/config/controller.go

```diff
@@ -0,0 +1,363 @@
+)
+
+const (
+	controllerName = "d8-module-config-controller"
+
+	moduleConfigFinalizer = "modules.deckhouse.io/module-config"
+
+	deleteReleasesAfter = 5 * time.Minute
+)
+
+func RegisterController(
+	runtimeManager manager.Manager,
+	handler *confighandler.Handler,
+	mm moduleManager,
+	ms *metricstorage.MetricStorage,
+	loader *moduleloader.Loader,
+	bundle string,
+) error {
+	r := &reconciler{
+		init:          &sync.WaitGroup{},
+		client:        runtimeManager.GetClient(),
+		log:           log.WithField("component", "ModuleConfigController"),
+		handler:       handler,
+		moduleManager: mm,
+		metricStorage: ms,
+		moduleLoader:  loader,
+		bundle:        bundle,
+	}
+
+	r.init.Add(1)
+
+	// sync module configs
+	if err := runtimeManager.Add(manager.RunnableFunc(r.syncModules)); err != nil {
+		return err
+	}
+
+	configController, err := controller.New(controllerName, runtimeManager, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(runtimeManager).
+		For(&v1alpha1.ModuleConfig{}).
+		WithEventFilter(predicate.GenerationChangedPredicate{}).
+		Complete(configController)
+}
+
+type reconciler struct {
+	init          *sync.WaitGroup
+	client        client.Client
+	log           logger.Logger
+	handler       *confighandler.Handler
+	moduleManager moduleManager
+	metricStorage *metricstorage.MetricStorage
+	moduleLoader  *moduleloader.Loader
+	bundle        string
+}
+
+type moduleManager interface {
+	AreModulesInited() bool
+	IsModuleEnabled(moduleName string) bool
+	GetModuleNames() []string
+	GetModuleEventsChannel() chan events.ModuleEvent
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// wait until init
+	r.init.Wait()
+
+	r.log.Debugf("reconciling the '%s' module config", req.Name)
+	moduleConfig := new(v1alpha1.ModuleConfig)
+	if err := r.client.Get(ctx, client.ObjectKey{Name: req.Name}, moduleConfig); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.log.Warnf("the '%s' module config not found", req.Name)
+			return ctrl.Result{}, nil
+		}
+		r.log.Errorf("failed to get the '%s' module config: %v", req.Name, err)
+		return ctrl.Result{}, err
+	}
+
+	// handle delete event
+	if !moduleConfig.DeletionTimestamp.IsZero() {
+		r.log.Debugf("deleting the '%s' module config", req.Name)
+		return r.deleteModuleConfig(ctx, moduleConfig)
+	}
+
+	return r.handleModuleConfig(ctx, moduleConfig)
+}
+
+func (r *reconciler) handleModuleConfig(ctx context.Context, moduleConfig *v1alpha1.ModuleConfig) (ctrl.Result, error) {
+	r.log.Debugf("handle the event for the module '%s' config", moduleConfig.Name)
+	r.handler.HandleEvent(ctx, moduleConfig, config.EventUpdate)
+
+	r.log.Debugf("refresh the '%s' module config status", moduleConfig.Name)
+	if err := r.refreshModuleConfigStatus(ctx, moduleConfig.Name); err != nil {
+		return ctrl.Result{Requeue: true}, err
```

Comment:
```suggestion
		return ctrl.Result{}, err
```
because if err != nil - Result will be ignored

Reply:
fixed